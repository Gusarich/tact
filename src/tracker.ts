import {
    Address,
    Cell,
    ComputeSkipReason,
    Contract,
    fromNano,
    Message,
    Slice,
    Transaction,
} from "@ton/core";
import { SendMessageResult } from "@ton/sandbox";

export type TrackedEvent =
    | {
          $type: "deploy";
      }
    | {
          $type: "frozen";
      }
    | {
          $type: "deleted";
      }
    | {
          $type: "received";
          message: TrackedMessage;
      }
    | {
          $type: "received-bounced";
          message: TrackedMessage;
      }
    | {
          $type: "failed";
          errorCode: number;
          errorMessage?: string;
      }
    | {
          $type: "processed";
          gasUsed: bigint;
      }
    | {
          $type: "skipped";
          reason: ComputeSkipReason;
      }
    | {
          $type: "sent";
          messages: TrackedMessage[];
      }
    | {
          $type: "sent-bounced";
          message: TrackedMessage;
      }
    | {
          $type: "sent-bounced-failed";
      }
    | {
          $type: "storage-charged";
          amount: string;
      };

export type TrackedBody =
    | {
          type: "empty";
      }
    | {
          type: "cell";
          cell: string;
      }
    | {
          type: "text";
          text: string;
      }
    | {
          type: "known";
          value: unknown;
      };

export type TrackedMessage =
    | {
          type: "external-in";
          to: string;
          body: TrackedBody;
      }
    | {
          type: "external-out";
          to: string | null;
          body: TrackedBody;
      }
    | {
          type: "internal";
          from: string;
          to: string;
          value: string;
          bounce: boolean;
          body: TrackedBody;
      };

interface ContractWithParser extends Contract {
    parseMessage(slice: Slice): unknown;
}

export class Tracker {
    private _contracts: Map<string, ContractWithParser> = new Map();
    private _events: TrackedEvent[][] = [];

    reset() {
        this._contracts = new Map();
        this._events = [];
    }

    track(contract: ContractWithParser) {
        this._contracts.set(contract.address.toRawString(), contract);
    }

    collect() {
        const events = this._events;
        this._events = [];
        return events;
    }

    parseBody(src: Cell, addresses: Address[]): TrackedBody {
        const sc = src.beginParse();

        // Empty case
        if (sc.remainingBits === 0 && sc.remainingRefs === 0) {
            return { type: "empty" };
        }

        // Too short for op
        if (sc.remainingBits <= 32) {
            return { type: "cell", cell: src.toString() };
        }

        // If text
        if (sc.preloadUint(32) === 0) {
            return { type: "text", text: sc.loadStringTail() };
        }

        // If known
        for (const a of addresses) {
            const contract = this._contracts.get(a.toRawString());
            if (contract) {
                try {
                    return {
                        type: "known",
                        value: contract.parseMessage(sc),
                    };
                } catch (e) {
                    /* empty */
                }
            }
        }

        // Fallback
        return { type: "cell", cell: src.toString() };
    }

    parseMessage(src: Message): TrackedMessage {
        // Internal message
        if (src.info.type === "internal") {
            const from = src.info.src.toString({ testOnly: true });
            const to = src.info.dest.toString({ testOnly: true });

            return {
                type: "internal",
                from,
                to,
                value: fromNano(src.info.value.coins),
                bounce: src.info.bounce,
                body: this.parseBody(src.body, [src.info.src, src.info.dest]),
            };
        }

        // External in
        if (src.info.type === "external-in") {
            const to = src.info.dest.toString({ testOnly: true });
            return {
                type: "external-in",
                to: to,
                body: this.parseBody(src.body, [src.info.dest]),
            };
        }

        // External out
        return {
            type: "external-out",
            to: src.info.dest ? src.info.dest.toString() : null,
            body: this.parseBody(src.body, [src.info.src]),
        };
    }

    parseTransaction(tx: Transaction) {
        // Some sanity checks
        if (!tx.inMessage) {
            throw Error("Tick-tock transaction is not supported");
        }
        if (tx.description.type !== "generic") {
            throw Error("Non-generic transaction is not supported");
        }

        const events: TrackedEvent[] = [];

        // Check if deployed
        if (
            (tx.oldStatus === "non-existing" ||
                tx.oldStatus === "uninitialized") &&
            tx.endStatus === "active"
        ) {
            events.push({ $type: "deploy" });
        }

        if (tx.description.storagePhase) {
            if (tx.description.storagePhase.storageFeesCollected > 0n) {
                events.push({
                    $type: "storage-charged",
                    amount: fromNano(
                        tx.description.storagePhase.storageFeesCollected,
                    ),
                });
            }
            if (tx.description.storagePhase.statusChange === "frozen") {
                events.push({ $type: "frozen" });
            }
            if (tx.description.storagePhase.statusChange === "deleted") {
                events.push({ $type: "deleted" });
            }
        }

        // Incoming message
        const msg = this.parseMessage(tx.inMessage);
        if (
            tx.inMessage.info.type === "internal" &&
            tx.inMessage.info.bounced
        ) {
            events.push({ $type: "received-bounced", message: msg });
        } else {
            events.push({ $type: "received", message: msg });
        }

        // Processing
        if (tx.description.computePhase.type === "vm") {
            if (tx.description.computePhase.success) {
                events.push({
                    $type: "processed",
                    gasUsed: tx.description.computePhase.gasUsed,
                });
            } else {
                events.push({
                    $type: "failed",
                    errorCode: tx.description.computePhase.exitCode,
                });
            }
        } else {
            events.push({
                $type: "skipped",
                reason: tx.description.computePhase.reason,
            });
        }

        // Outgoing messages
        for (const outgoingMessage of tx.outMessages.values()) {
            const msg = this.parseMessage(outgoingMessage);
            if (
                outgoingMessage.info.type === "internal" &&
                outgoingMessage.info.bounced
            ) {
                events.push({ $type: "sent-bounced", message: msg });
            } else {
                events.push({ $type: "sent", messages: [msg] });
            }
        }

        // Persist events
        this._events.push(events);
    }

    parse(result: SendMessageResult) {
        for (const tx of result.transactions) {
            this.parseTransaction(tx);
        }
    }
}
