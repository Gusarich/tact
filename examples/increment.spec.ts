import { IncrementContract } from "./output/increment_IncrementContract";
import { Blockchain } from "@ton/sandbox";
import { toNano } from "@ton/core";
import { Tracker } from "../src/tracker";

describe("increment", () => {
    it("should deploy", async () => {
        // Create wallet
        const blockchain = await Blockchain.create();
        const treasure = await blockchain.treasury("treasure");
        const contract = blockchain.openContract(
            await IncrementContract.fromInit(),
        );
        const tracker = new Tracker();
        tracker.track(contract);

        tracker.parse(
            await contract.send(
                treasure.getSender(),
                { value: toNano("10") },
                { $$type: "Deploy", queryId: 0n },
            ),
        );
        expect(tracker.collect()).toMatchSnapshot();

        // Send internal message
        tracker.parse(
            await contract.send(
                treasure.getSender(),
                { value: toNano("10") },
                { $$type: "Increment", key: 0n, value: -1232n },
            ),
        );
        expect(tracker.collect()).toMatchSnapshot();
    });
});
