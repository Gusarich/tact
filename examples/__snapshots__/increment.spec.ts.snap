// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`increment should deploy 1`] = `
[
  [
    {
      "$type": "received",
      "message": {
        "body": {
          "cell": "x{3FC5DCD27049199660E33C40BFCC6DC7FB7EF8A985CDDF3FE82138BDAA67B96FC_}
 x{D001}
  x{62004CD5D51F4E90E1518BE466E8404BBBE1BC7D7F162309FCAA92D4DA1CD35DF593A812A05F2000000000000000000000000000023251AA62D80000000000000002_}
   x{FF00F4A413F4BCF2C80B}
    x{62_}
     x{D001D0D3030171B0A301FA400120D74981010BBAF2E08820D70B0A208104FFBAF2D0898309BAF2E088545053036F04F86102F862DB3C5514DB3CF2E082C8F84301CC7F01CA0055405045F40002C8F400F40012F40012F400C901CCC9ED54}
      x{ED44D0D401F863D200018E12F404D401D0F404F404F404F4043010456C15E030F828D70B0A8309BAF2E089DB3C}
       x{6D6D6D6D6D}
      x{0192307FE07021D749C21F953020D70B1FDE20821020064F3BBA8E5730D31F01821020064F3BBAF2E081810101D700810101D700596C1281010120103841805280216E955B59F45A3098C801CF004133F442E281010BF84210344107810101216E955B59F4593098C801CF004133F441E2017FE02082104077D4C8}
       x{BA8E6D30D31F0182104077D4C8BAF2E081810101D70001312481010122714133F40C6FA19401D70030925B6DE2206E8E1A3014810101017F71216E955B59F45A3098C801CF004133F442E28E2081010101206EF2D080B310361271216E955B59F45A3098C801CF004133F442E2E2037FE0208210E29D0FAABAE30220}
        x{30D31F018210E29D0FAABAF2E081810101D700D2000191D4926D01E2596C128200CE29258101012459F40D6FA192306DDF6EF2F4102481010159206E953059F45A30944133F415E2027F}
        x{82106668EFB2BA8E9830D31F0182106668EFB2BAF2E081810101D7000131DB3C7FE08210946A98B6BA8EA7D31F018210946A98B6BAF2E081D33F0131C8018210AFF90F5758CB1FCB3FC9F84201706DDB3C7FE03070}
         x{8101016D5312104959216E955B59F45A3098C801CF004133F442E204810101266D71216E955B59F45A3098C801CF004133F442E203810101266D206E953059F45A30944133F415E281010BF84210246D810101216E955B59F4593098C801CF004133F441E28101016D}
          x{206E92306D8E10206EF2D0806F21C80101810101CF00C9E24170206E953059F45A30944133F415E210344130}
         x{6D6D226EB3995B206EF2D0806F22019132E2102470030480425023DB3C}
          x{C87101CA01500701CA007001CA02500520D74981010BBAF2E08820D70B0A208104FFBAF2D0898309BAF2E088CF165003FA027001CA68236EB3917F93246EB3E2973333017001CA00E30D216EB39C7F01CA0001206EF2D08001CC95317001CA00E2C901FB00}
           x{7F01CA00C87001CA007001CA00246EB39D7F01CA0004206EF2D0805004CC9634037001CA00E2246EB39D7F01CA0004206EF2D0805004CC9634037001CA00E27001CA00027F01CA0002C958CC}
     x{2_}
      x{2_}
       x{BAF02DB3CDB3C6C51}
        x{ED44D0D401F863D200018E12F404D401D0F404F404F404F4043010456C15E030F828D70B0A8309BAF2E089DB3C}
         x{6D6D6D6D6D}
        x{21}
       x{BB407DB3CDB3C6C51}
        x{ED44D0D401F863D200018E12F404D401D0F404F404F404F4043010456C15E030F828D70B0A8309BAF2E089DB3C}
         x{6D6D6D6D6D}
        x{24}
      x{BE15F76A268690000C_}
   x{4_}
    x{C_}
     x{A17443_}
      x{FF00F4A413F4BCF2C80B}
       x{62_}
        x{D001D0D3030171B0A301FA400120D74981010BBAF2E08820D70B0A208104FFBAF2D0898309BAF2E088545053036F04F86102F862DB3C5514DB3CF2E082C8F84301CC7F01CA0055405045F40002C8F400F40012F40012F400C901CCC9ED54}
         x{ED44D0D401F863D200018E12F404D401D0F404F404F404F4043010456C15E030F828D70B0A8309BAF2E089DB3C}
          x{6D6D6D6D6D}
         x{0192307FE07021D749C21F953020D70B1FDE20821020064F3BBA8E5730D31F01821020064F3BBAF2E081810101D700810101D700596C1281010120103841805280216E955B59F45A3098C801CF004133F442E281010BF84210344107810101216E955B59F4593098C801CF004133F441E2017FE02082104077D4C8}
          x{BA8E6D30D31F0182104077D4C8BAF2E081810101D70001312481010122714133F40C6FA19401D70030925B6DE2206E8E1A3014810101017F71216E955B59F45A3098C801CF004133F442E28E2081010101206EF2D080B310361271216E955B59F45A3098C801CF004133F442E2E2037FE0208210E29D0FAABAE30220}
           x{30D31F018210E29D0FAABAF2E081810101D700D2000191D4926D01E2596C128200CE29258101012459F40D6FA192306DDF6EF2F4102481010159206E953059F45A30944133F415E2027F}
           x{82106668EFB2BA8E9830D31F0182106668EFB2BAF2E081810101D7000131DB3C7FE08210946A98B6BA8EA7D31F018210946A98B6BAF2E081D33F0131C8018210AFF90F5758CB1FCB3FC9F84201706DDB3C7FE03070}
            x{8101016D5312104959216E955B59F45A3098C801CF004133F442E204810101266D71216E955B59F45A3098C801CF004133F442E203810101266D206E953059F45A30944133F415E281010BF84210246D810101216E955B59F4593098C801CF004133F441E28101016D}
             x{206E92306D8E10206EF2D0806F21C80101810101CF00C9E24170206E953059F45A30944133F415E210344130}
            x{6D6D226EB3995B206EF2D0806F22019132E2102470030480425023DB3C}
             x{C87101CA01500701CA007001CA02500520D74981010BBAF2E08820D70B0A208104FFBAF2D0898309BAF2E088CF165003FA027001CA68236EB3917F93246EB3E2973333017001CA00E30D216EB39C7F01CA0001206EF2D08001CC95317001CA00E2C901FB00}
              x{7F01CA00C87001CA007001CA00246EB39D7F01CA0004206EF2D0805004CC9634037001CA00E2246EB39D7F01CA0004206EF2D0805004CC9634037001CA00E27001CA00027F01CA0002C958CC}
        x{2_}
         x{2_}
          x{BAF02DB3CDB3C6C51}
           x{ED44D0D401F863D200018E12F404D401D0F404F404F404F4043010456C15E030F828D70B0A8309BAF2E089DB3C}
            x{6D6D6D6D6D}
           x{21}
          x{BB407DB3CDB3C6C51}
           x{ED44D0D401F863D200018E12F404D401D0F404F404F404F4043010456C15E030F828D70B0A8309BAF2E089DB3C}
            x{6D6D6D6D6D}
           x{24}
         x{BE15F76A268690000C_}",
          "type": "cell",
        },
        "to": "kQDd_g_2RL1F5eZ5BjlI7F56Ls2b-61XEl23HlO_fzJN8VhW",
        "type": "external-in",
      },
    },
    {
      "$type": "processed",
      "gasUsed": 1937n,
    },
    {
      "$type": "sent",
      "messages": [
        {
          "body": {
            "type": "known",
            "value": {
              "$$type": "Deploy",
              "queryId": 0n,
            },
          },
          "bounce": true,
          "from": "kQDd_g_2RL1F5eZ5BjlI7F56Ls2b-61XEl23HlO_fzJN8VhW",
          "to": "kQCZq6o-nSHCoxfIzdCAl3fDePr-LEYT-VUlqbQ5prvrJ7Y2",
          "type": "internal",
          "value": "10",
        },
      ],
    },
  ],
  [
    {
      "$type": "deploy",
    },
    {
      "$type": "received",
      "message": {
        "body": {
          "type": "known",
          "value": {
            "$$type": "Deploy",
            "queryId": 0n,
          },
        },
        "bounce": true,
        "from": "kQDd_g_2RL1F5eZ5BjlI7F56Ls2b-61XEl23HlO_fzJN8VhW",
        "to": "kQCZq6o-nSHCoxfIzdCAl3fDePr-LEYT-VUlqbQ5prvrJ7Y2",
        "type": "internal",
        "value": "10",
      },
    },
    {
      "$type": "processed",
      "gasUsed": 8473n,
    },
    {
      "$type": "sent",
      "messages": [
        {
          "body": {
            "type": "known",
            "value": {
              "$$type": "DeployOk",
              "queryId": 0n,
            },
          },
          "bounce": false,
          "from": "kQCZq6o-nSHCoxfIzdCAl3fDePr-LEYT-VUlqbQ5prvrJ7Y2",
          "to": "kQDd_g_2RL1F5eZ5BjlI7F56Ls2b-61XEl23HlO_fzJN8VhW",
          "type": "internal",
          "value": "9.9961324",
        },
      ],
    },
  ],
  [
    {
      "$type": "received",
      "message": {
        "body": {
          "type": "known",
          "value": {
            "$$type": "DeployOk",
            "queryId": 0n,
          },
        },
        "bounce": false,
        "from": "kQCZq6o-nSHCoxfIzdCAl3fDePr-LEYT-VUlqbQ5prvrJ7Y2",
        "to": "kQDd_g_2RL1F5eZ5BjlI7F56Ls2b-61XEl23HlO_fzJN8VhW",
        "type": "internal",
        "value": "9.9961324",
      },
    },
    {
      "$type": "processed",
      "gasUsed": 309n,
    },
  ],
]
`;

exports[`increment should deploy 2`] = `
[
  [
    {
      "$type": "received",
      "message": {
        "body": {
          "cell": "x{3FC5DCD27049199660E33C40BFCC6DC7FB7EF8A985CDDF3FE82138BDAA67B96FC_}
 x{D001}
  x{62004CD5D51F4E90E1518BE466E8404BBBE1BC7D7F162309FCAA92D4DA1CD35DF593A812A05F20000000000000000000000000000020064F3B00000000000000000000000000000000000000000000000000000000000000007FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFECC2_}",
          "type": "cell",
        },
        "to": "kQDd_g_2RL1F5eZ5BjlI7F56Ls2b-61XEl23HlO_fzJN8VhW",
        "type": "external-in",
      },
    },
    {
      "$type": "processed",
      "gasUsed": 1937n,
    },
    {
      "$type": "sent",
      "messages": [
        {
          "body": {
            "type": "known",
            "value": {
              "$$type": "Increment",
              "key": 0n,
              "value": -1232n,
            },
          },
          "bounce": true,
          "from": "kQDd_g_2RL1F5eZ5BjlI7F56Ls2b-61XEl23HlO_fzJN8VhW",
          "to": "kQCZq6o-nSHCoxfIzdCAl3fDePr-LEYT-VUlqbQ5prvrJ7Y2",
          "type": "internal",
          "value": "10",
        },
      ],
    },
  ],
  [
    {
      "$type": "received",
      "message": {
        "body": {
          "type": "known",
          "value": {
            "$$type": "Increment",
            "key": 0n,
            "value": -1232n,
          },
        },
        "bounce": true,
        "from": "kQDd_g_2RL1F5eZ5BjlI7F56Ls2b-61XEl23HlO_fzJN8VhW",
        "to": "kQCZq6o-nSHCoxfIzdCAl3fDePr-LEYT-VUlqbQ5prvrJ7Y2",
        "type": "internal",
        "value": "10",
      },
    },
    {
      "$type": "processed",
      "gasUsed": 5963n,
    },
  ],
]
`;
